### PR TITLE
Add 0.3 upgrade guide

### PR DESCRIPTION
**Pull request by ThetaSinner**

### CODE REVIEW

File: src/pages/_data/navigation/mainNav.json5 [hunk 1]
```diff
@@ -8,7 +8,8 @@
         { title: "Dev Tools Setup Tips", url: "/get-started/install-advanced/" },
         { title: "Setup Without Nix", url: "/get-started/install-without-nix/" },
         { title: "Distribute your App", url: "/get-started/distribute-your-app/" },
-        { title: "Holochain Upgrade 0.1 → 0.2", url: "/get-started/upgrade-holochain/" },
+        { title: "Holochain Upgrade 0.1 → 0.2", url: "/get-started/upgrade-holochain-0.2/" },
```
Comment (ThetaSinner) [inline src/pages/_data/navigation/mainNav.json5:11]: Is it okay to rename these? It will break existing links but it'd be nice to have a pattern for these file names going forward. If we do want to maintain links then maybe we could always have the link with no version suffix go to the latest guide?

File: src/pages/get-started/upgrade-holochain-0.3.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+To this:
+
+```js
+await client.appInfo();
+```
+
+This is because app client connections are now bound to the app that they are connected to. So there's no need to tell
+the conductor which app you want to operate on, it already knows.
+
+#### Update `@holochain/hc-spin`
```
Comment (ThetaSinner) [inline src/pages/get-started/upgrade-holochain-0.3.md:143]: Is there already information somewhere about dealing with electron security stuff on new OSes? I've upgraded my Ubuntu now and ran into this. 

I fixed it with 

```
sudo chown root /tmp/hello-world/node_modules/electron/dist/chrome-sandbox
sudo chmod 4755 /tmp/hello-world/node_modules/electron/dist/chrome-sandbox
```

If there's some general advice already been posted somewhere then I'd rather link that here!

Comment (matthme) [reply src/pages/get-started/upgrade-holochain-0.3.md:143]: Yes, @pdaoust added some with [this PR](https://github.com/holochain/docs-pages/pull/452)

Comment (matthme) [reply src/pages/get-started/upgrade-holochain-0.3.md:143]: So `..install-advanced#fixing-the-suid-sandbox-error-in-ubuntu-24-04` would probably the link to point to

**Review by matthme — APPROVED** [review]
Looking great! The only thing I'm wondering is whether this valuable guide deserves a more obvious link, potentially even somewhere on the main page of the dev portal.

**Review by c12i — APPROVED** [review]
Looks good to me

File: src/pages/get-started/upgrade-holochain-0.3.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+Updating the `flake.nix` won't make Nix use the new Holochain version when you next open a dev shell with `nix develop`. For the changes
+to take effect you need to run:
+
+```shell
+nix flake update
+```
+
+!!! note
+This takes all updates from Holonix, possibly including an update to the Rust version that Holonix is currently using. 
+You should run this command whenever you update your Cargo dependencies to the latest HDI/HDK versions.
```
Comment (jost-s) [inline src/pages/get-started/upgrade-holochain-0.3.md:67]: What I do in cargo.tomls has no effect on the flake or its lock. This suggests to run `nix flake update` every time I update HDI/HDK. What's the reason for that?

File: src/pages/get-started/upgrade-holochain-0.3.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+!!! note
+The following steps only apply to projects that have been either generated by the Holochain scaffolding tool
+or have followed the same folder layout. Otherwise, you will need to adjust the commands.
+!!!
+
+#### Update the `@holochain/client`
+
+In your project's `ui/package.json` and `tests/package.json` you will find a dependency on `@holochain/client`. These should be using `0.16.x`, which is the version compatible with Holochain 0.2. To work with Holochain 0.3, you need to update to `0.17.x`.
+
+You can always check what versions of the Holochain client are available on the [NPM registry](https://www.npmjs.com/package/@holochain/client?activeTab=versions) and you can find out what versions are compatible with a Holochain version by checking the [project's documentation](https://github.com/holochain/holochain-client-js).
```
Comment (jost-s) [inline src/pages/get-started/upgrade-holochain-0.3.md:97]:
```suggestion
You can always check what versions of the Holochain client are available on the [NPM registry](https://www.npmjs.com/package/@holochain/client?activeTab=versions) and you can find out what versions are compatible with a Holochain version by checking the [project's documentation](https://github.com/holochain/holochain-client-js?tab=readme-ov-file#installation).
```

File: src/pages/get-started/upgrade-holochain-0.3.md [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+npm install --workspace tests @holochain/tryorama@^0.16.0
+```
+
+!!!note
+This will not install version `0.16.0` if a newer version has been released. Check your `package-lock.json` to see exactly what version got installed. It will be `0.16.x` because the `^` constraint only permits patch upgrades, so you will get a version that's compatible with Holochain 0.3.
+!!!
+
+### Update Cargo dependencies
+
+This section is harder to write a general guide for because it's common for hApps to add dependencies on other Holochain crates. If you have added other dependencies than the `hdi` and `hdk` to your project then you will need to update those too but figuring out which versions you
```
Comment (jost-s) [inline src/pages/get-started/upgrade-holochain-0.3.md:172]:
```suggestion
This section is harder to write a general guide for, because it's common for hApps to add dependencies on other Holochain crates. If you have added other dependencies than the `hdi` and `hdk` to your project, then you will need to update those too, but figuring out which versions you
```

**Review by jost-s — APPROVED** [review]

File: src/pages/get-started/upgrade-holochain-0.3.md (hunk 2 above)
Comment (ThetaSinner) [reply src/pages/get-started/upgrade-holochain-0.3.md:143]: Super, thank you!

File: src/pages/get-started/upgrade-holochain-0.3.md (hunk 3 above)
Comment (ThetaSinner) [reply src/pages/get-started/upgrade-holochain-0.3.md:67]: They're not directly related but the Nix updates should be done at least as frequently. I'll reword this to make it clearer what I'm getting at

File: src/pages/get-started/upgrade-holochain-0.3.md (hunk 5 above)
Comment (ThetaSinner) [reply src/pages/get-started/upgrade-holochain-0.3.md]: You don't like my never-ending sentences? :) 

Thank you!

**Comment by ThetaSinner** [comment]
> The only thing I'm wondering is whether this valuable guide deserves a more obvious link, potentially even somewhere on the main page of the dev portal.

I always think of getting started as the main page because I tend to go there ->

image

But that's a good point, I wonder where else it might be good to put this

**Comment by matthme** [comment]
I did not necessarily mean to remove it from there but to add an additional link to it somewhere else where it's more obvious. Like on the main page something along the lines of (once 0.3 is marked as the recommended version): "Holochain 0.3 is now the recommended version. If you've been using 0.2 here is a guide on how to upgrade your happ from 0.2 to 0.3".

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+Updating the `flake.nix` won't make Nix use the new Holochain version when you next open a dev shell with `nix develop`. For the changes
+to take effect you need to run:
+
+```shell
+nix flake update
+```
+
+!!! note
+This takes all updates from Holonix, possibly including an update to the Rust version that Holonix is currently using.
```
Comment (jost-s) [inline src/pages/upgrade/upgrade-holochain-0.3.md:67]: Ah, you meant so that all crates can be pulled from the cache.

Comment (ThetaSinner) [reply src/pages/upgrade/upgrade-holochain-0.3.md:67]: Well it can't fetch Cargo deps from the cache pre-built, but it will mean that your CI won't start rebuilding all of Holochain on every run. So it saves downloading and building all of Holochain's deps to assemble the binary to test against yes

Comment (ThetaSinner) [reply src/pages/upgrade/upgrade-holochain-0.3.md:67]: Same thing if you switch machine or clear your Nix cache, you'll be able to fetch what you need rather than spending 45 minutes compiling everything again. Which is probably on the low end, it takes a couple of hours on a laptop.

**Comment by ThetaSinner** [comment]
@matthme How does that look? The only minor issue is that I've re-used an icon which looks a bit off on the main page. I'm not sure where those are sourced from because they're embedded through the SCSS for the site.

**Comment by matthme** [comment]
Looks great :+1:

**Comment by ThetaSinner** [comment]
Super, thank you! :)

File: src/pages/_data/navigation/mainNav.json5 (hunk 1 above)
Comment (pdaoust) [reply src/pages/_data/navigation/mainNav.json5]: yup, totally a good idea. The only caveat is that we should always add a 301 redirect to `/netlify.toml` when we do this, for search results and bookmarks.

File: src/pages/upgrade/upgrade-holochain-0.3.md (hunk 6 above)
Comment (pdaoust) [reply src/pages/upgrade/upgrade-holochain-0.3.md:67]: laptop over here 👋🏼  even just compiling `hc-scaffold` takes a few hours, but you will not wrest my 7yo ThinkPad from my grasp!

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+#### General advice for upgrading
+
+It's likely that you will need to make more changes than have been mentioned above, but hopefully this guide has given you a head start! Here are some other things to look out for in this upgrade:
+- There have been significant changes to Holochain's data validation. You may see validation failures that you didn't see before upgrading. Our experience of helping people upgrade to dev versions of 0.3 is that these are often due to issues in your app's validation logic. 
+  Please check your validation logic and check whether there might be an issue there. If not, then please reach out on Discord or open a [Github issue](https://github.com/holochain/holochain/issues/new/choose) for help.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:291]:
```suggestion
  Please check your validation logic and check whether there might be an issue there. If not, then please reach out on [Discord](https://discord.gg/E3EXyPBj3F) or open a [Github issue](https://github.com/holochain/holochain/issues/new/choose) for help.
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+- There have been changes to the data structures provided for app validation. This means that you will need to update your validation code to remove some fields that are no longer available. You should check the changelog [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#hdi-040-beta-dev36).
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:292]:
```suggestion
* There have been changes to the data structures provided for app validation. This means that you will need to update your validation code to remove some fields that are no longer available. You should check the changelog [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#hdi-040-beta-dev36).
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+  If you need to retain the fields that have been removed for your validation logic, then your best option is probably to scaffold an app for 0.3 and take a look at the validation code that is generated.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:293]: What would be the outcome of this recommendation? FWIU the changes are mainly that previous-action/entry deps for CRUD ops aren't fetched, and instead you're just given a hash. Can we say something about that here?

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+  You can do this by running `nix run --override-input versions "github:holochain/holochain?dir=versions/0_3" github:holochain/holochain#hc-scaffold`.
+- There have been significant app interface changes for Holochain 0.3. There is a changelog entry [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#holochain-030-beta-dev47). Most problems will be taken care of by picking
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:295]:
```suggestion
* There have been significant WebSocket app interface changes for Holochain 0.3. There is a changelog entry [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#holochain-030-beta-dev47). Most problems will be taken care of by picking
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+  the right client versions and following any changes that are required to use the new client. However, if you are integrating with Holochain's app websocket directly or maintaining your own client then you should get in touch if you run into any difficulties upgrading.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:296]:
```suggestion
  the right client versions and following any changes that are required to use the new client. However, if you are integrating with Holochain's app interface directly or maintaining your own client then you should get in touch if you run into any difficulties upgrading.
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+- Databases up to 0.3 have been unencrypted by default. From 0.3 onwards, databases are always encrypted by default. You can disable this by building Holochain with different options if you need to, or we can show you how to open the encrypted databases if you need.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:297]:
```suggestion
* Databases prior to 0.3 have been unencrypted by default. From 0.3 onwards, databases are always encrypted by default. You can disable this by building Holochain with different options if you need to, or we can show you how to open the encrypted databases.
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+  For now, we are assuming that people don't often need to do this so there is no guide. If this assumption turns out to be wrong, we can create a guide for this!
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:298]:
```suggestion
  For now, we are assuming that people don't often need to do this, so there is no guide. If this assumption turns out to be wrong, we can create one!
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+- A change was made to help hApp developers spot mistakes in DNA manifests that might catch you out if you have extra fields or old fields in your DNA manifest. You can read about this change [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#holochain_types-030-beta-dev24).
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:299]:
```suggestion
* A change was made to help hApp developers spot mistakes in DNA manifests that might catch you out if they contain extra fields or obsolete fields. You can read about this change [here](https://github.com/holochain/holochain/blob/develop-0.3/CHANGELOG.md#holochain_types-030-beta-dev24).
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+- Holochain 0.2 and 0.3 are not network compatible. That means that if you have users of you hApp still running Holochain 0.2, then even with the same DNA hash, they would not be able to connect to users running Holochain 0.3.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:300]:
```suggestion
* Holochain 0.2 and 0.3 are not network-compatible. That means that if you have users of your hApp still running Holochain 0.2, then even with the same DNA hash, they would not be able to connect to users running Holochain 0.3.
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+  The changes between 0.2 and 0.3 are not as significant as they were between 0.1 and 0.2, but they still prevent communication between the two versions.
+
+### Next steps
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:303]:
```suggestion
## Next steps
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+
+You should check that your project builds, tests are passing and that anything else your CI will check is working. You should also check that your hApp is still working because static checks and tests might not catch changes that are required in your UI.
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:305]:
```suggestion
You should check that your project builds, tests are passing, and that anything else your CI will check is working. You should also check that your hApp is still working because static checks and tests might not catch changes that are required in your UI.
```

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+For general project health, you should check if any other dependencies in your project need an update. Please refer to NPM and Cargo documentation for this.
+
+That should be everything, you can now develop against Holochain 0.3! Good luck, and if you need help then please do get in touch on Discord or open a [Github issue](https://github.com/holochain/holochain/issues/new/choose).
```
Comment (pdaoust) [inline src/pages/upgrade/upgrade-holochain-0.3.md:309]:
```suggestion
That should be everything --- you can now develop against Holochain 0.3! Good luck, and once again, if you need help then please do get in touch on [Discord](https://discord.gg/E3EXyPBj3F) or open a [Github issue](https://github.com/holochain/holochain/issues/new/choose).
```

**Review by pdaoust — DISMISSED** [review]
Did a formatting/grammar/etc sweep; mostly tiny changes. There are a few questions in there though. I'm also going to follow up with a PR on your PR to get the IA sorted out -- I want to avoid creating another category; maybe I'll put it under Resources.

File: src/pages/upgrade/upgrade-holochain-0.3.md (hunk 6 above)
Comment (ThetaSinner) [reply src/pages/upgrade/upgrade-holochain-0.3.md:67]: It's weird with computers, I love mine until something much faster comes along then suddenly all my feelings transfer to the faster machine :D

File: src/pages/upgrade/upgrade-holochain-0.3.md (hunk 9 above)
Comment (ThetaSinner) [reply src/pages/upgrade/upgrade-holochain-0.3.md:293]: The way the scaffolding was updated, it effectively maintains the existing way of doing things. That means that getting back the removed data can be achieved by merging in logic from the scaffolded `lib.rs` in the generated integrity zome.

I think the changelog describes this better than I can. Do we want to move something from the changelog to here or is linking to it enough?

File: src/pages/_data/navigation/mainNav.json5 (hunk 1 above)
Comment (ThetaSinner) [reply src/pages/_data/navigation/mainNav.json5]: Added the 301. The 302 I changed my mind about, because we now have a top level page that lists all the version migration guides so there's an easy place to send people to look for the right guide

File: src/pages/upgrade/upgrade-holochain-0.3.md [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+```rust
+get(r.action().entry_hash().unwrap().clone(), GetOptions::content())?;
+```
+
+Which should be updated to:
+```rust
+get(r.action().entry_hash().unwrap().clone(), GetOptions::network())?;
+```
+
+The `Local` variant is intended to support a local first app which is willing to operate with the currently available data.
```
[the remark this answers is not in the document]
Comment (ThetaSinner) [reply src/pages/upgrade/upgrade-holochain-0.3.md:293]: I wasn't that keen on the use of 'stale' here, so I tweaked this a bit. I think stale  depends on context. Using 2005's market data for trading in 2024 would be a serious case of stale data. Choosing to work offline and syncing with other people when you next have a network connection to them could be a choice. It could be a requirement though, if you live somewhere with an unreliable connection or use mobile data during the day which is expensive to sync over. To me these aren't always about stale data, it's more about having the flexibility to use your data without connectivity.

File: netlify.toml [hunk 20]
```diff
@@ -130,4 +130,8 @@
     to = "/get-started/"
     status = 301
 
-# This comment does absolutely nothing and can be deleted.
```
Comment (ThetaSinner) [inline netlify.toml:133]: Okay then :)

**Comment by ThetaSinner** [comment]
Github has got a bit confused about whitespace changes but I think I've resolved all the comments now

File: src/pages/upgrade/upgrade-holochain-0.3.md (hunk 9 above)
Comment (pdaoust) [reply src/pages/upgrade/upgrade-holochain-0.3.md:303]: how about something like this for the last sentence? "scaffold an app for 0.3 and take a look at _how its validation code retrieves the data that's no longer part of the data structures._"

File: netlify.toml (hunk 20 above)
Comment (pdaoust) [reply netlify.toml:133]: ha ha, I think I was probably trying to force a CI run or bust a cache or something!

**Comment by pdaoust** [comment]
I've added my final review in #456. Please review and merge @ThetaSinner , and it'll go into this branch, and then you can merge this one into `main` without waiting for me to approve it.
